[ux] Created new navigation menu

niteshsinha17 opened the pull request:
A new Navigation Menu is created in this PR. Its backend is being created here. https://github.com/openwisp/openwisp-utils/pull/175
This branch contains only design. 

@nemesisdesign and @atb00ker please check out this one.

![menu](https://user-images.githubusercontent.com/42765292/117002590-5f471a80-ad01-11eb-84a5-b6fe609a67df.gif)

Design samples available here. https://miro.com/app/board/o9J_lMAQQZk=/
I request other community members to check this once and give their valuable views about it. Thank you

closes #188

atb00ker commented:
Looks great, couple of UI feedbacks:
1. Top menu can be thinner, it's looks too fat (The top section in drawer and the rest of the page should be of different width!)
2. (nitpick) The grey color of opened accordion makes it look like it's disabled.
3. @nemesisdesign , I think we need to think of change management a bit, I was trying to click on "admin" on the right-top side of the screen to logout when I realized it has moved to be inside left side menu, maybe we can open the drawer if the user clicks on the name of the admin. :smile:

nemesisdesign reviewed (commented):
@niteshsinha17 great progress! 

I think the light variant in your proposal looks good so can you try that?
Do you know where to find the logos? https://openwisp.io/docs/general/press.html#logos-and-graphic-material

Regarding what @atb00ker wrote:

> Looks great, couple of UI feedbacks:
> 
> 1. Top menu can be thinner, it's looks too fat (The top section in drawer and the rest of the page should be of different width!)

The top bar you mean? I agree, it could have a slightly smaller height I believe.

> 2. (nitpick) The grey color of opened accordion makes it look like it's disabled.

Yes we'll have to find something better but right now I can't think of something, let's note it down so we improve it soon.

> 3. @nemesisdesign , I think we need to think of change management a bit, I was trying to click on "admin" on the right-top side of the screen to logout when I realized it has moved to be inside left side menu, maybe we can open the drawer if the user clicks on the name of the admin. smile

May be a good idea, what do you think @niteshsinha17?

> P.S: CSS file has new conflicts, you may need to look into that again.

:+1:

niteshsinha17 commented:
Thanks @atb00ker and @nemesisdesign 
>I think the light variant in your proposal looks good so can you try that?

I will make the same design.

>1. Top menu can be thinner, it's looks too fat (The top section in drawer and the rest of the page should be of different width!)

For this, we can reduce its height. 

>Yes we'll have to find something better but right now I can't think of something, let's note it down so we improve it soon.

I think, let us first create the functions that are needed. We can change the color scheme anytime before ending the work.

>@nemesisdesign , I think we need to think of change management a bit, I was trying to click on "admin" on the right-top side of the screen to logout when I realized it has moved to be inside left side menu, maybe we can open the drawer if the user clicks on the name of the admin. smile

I agree, We can do this.

>P.S: CSS file has new conflicts, you may need to look into that again.

I will fix it soon 😄 

I want to know something. Should I start designing filters it this particular PR?

niteshsinha17 commented:
@atb00ker and @nemesisdesign Its ready now.
Tests are failing because of using static data right now.

nemesisdesign commented:
> @atb00ker and @nemesisdesign Its ready now.
> Tests are failing because of using static data right now.
> 
> Do I need to change styling of other things like button and filter in this PR?

@niteshsinha17 no, we'll work on those in separate patches. We will start pushing these changes to a new branch.

Question, what is the relation of this PR with #175? Does it depend on it?

niteshsinha17 commented:
>Question, what is the relation of this PR with #175? Does it depend on it?

Data is going to come from the changes made in that PR.

atb00ker commented:
> 1. Top menu can be thinner, it's looks too fat (The top section in drawer and the rest of the page should be of different width!)
For this, we can reduce its height.

So... can you please do that? Make `#header` element as `60px` or something like that.

atb00ker reviewed (commented):
I think this is close to read, 
found one more problem: The scrollbars, we only need the vertical one, that too preferably should be visible only when required.
![ueanu9F](https://user-images.githubusercontent.com/24985760/117337443-fdcaab80-aeba-11eb-80c9-636cf90e79dd.png)

Also, can you please add a lot of dummy data and test the scrollbar looks good! :smile:

niteshsinha17 commented:
@atb00ker  I have found some other issues too like colour of group name and menu was not looking good when we use long names. I have fixed them all. Also increased spacing between links. Also added some more dummy data.
>I think this is close to read,

Were you talking about spacing between links and space between name and icon? 

Does it looks good?
![image](https://user-images.githubusercontent.com/42765292/117491264-97628d80-af8d-11eb-99cd-14cc4c666b87.png)

atb00ker commented:
When this is merged, let's also look at the scrollbars and make them modern to make them feel the same as the new design:
We will need to test them with Chrome & Firefox!
Please pick any design you like here is just a sample list: https://www.cssscript.com/best-custom-scrollbar-javascript-libraries/ 

![scrollbar1](https://user-images.githubusercontent.com/24985760/117536333-14d2e000-b018-11eb-8486-92117fc3a24c.png)
![scrollbar2](https://user-images.githubusercontent.com/24985760/117536336-169ca380-b018-11eb-846a-1b9f8f178b59.png)

atb00ker commented:
P.S: @niteshsinha17 Which browser do you test the changes on (I'll test on something else so that every single change is tested on multiple browser)?

atb00ker reviewed (commented):
Oh, I just noticed there is no cross button in the menu!
While click anywhere outside works, we also need a cross button for visually impaired people!
Somewhere at the top in the menu preferably!

Talking about accessibility -- keyboard control in the menu should also work. (low priority)
Also, I noticed another minor low priority issue we would tackle in the end:
![anotherIssue](https://user-images.githubusercontent.com/24985760/117536819-ef93a100-b01a-11eb-9a23-5e6b1567fd0c.png)

Please note these low priority issues somewhere, we would have come back to them in the future! :smile:
(We can make a GitHub issue and track everything about the project there!)

niteshsinha17 commented:
> P.S: @niteshsinha17 Which browser do yo### u test the changes on? (I'll test on something else so that every single change is tested on multiple browser)

I test it on both Edge and chrome. I am also compiling css file with sass compiler so it adds some extra css properties for browser support. I think I had read somewhere about this but don't remember where right now. Can you please confirm is it good or not? 
Also, I will make the changes soon.

nemesisdesign reviewed (commented):
This is what I meant for light variant:

![Screenshot from 2021-05-09 19-05-34](https://user-images.githubusercontent.com/841044/117591307-9d757d00-b0f9-11eb-8f49-9010d38f73a4.png)

niteshsinha17 commented:
> This is what I meant for light variant. 

Oh, I will change it 👍

niteshsinha17 commented:
@nemesisdesign should we change background of logo present in login page. What do you think?

nemesisdesign commented:
> @nemesisdesign should we change background of logo present in login page. What do you think?

yep :+1:

niteshsinha17 commented:
@nemesisdesign Its ready now

atb00ker commented:
> We can handle them in a new PR.

Cool, but do note them down! :smile:
The rest looks good! :smile:

niteshsinha17 commented:
> > We can handle them in a new PR.
> 
> Cool, but do note them down! 😄
> The rest looks good! 😄

Sure I will open an issue for it.

atb00ker reviewed (changes requested):
I found a problem, cross button is missing!
![cross-button-is-missing](https://user-images.githubusercontent.com/24985760/118705587-0b2a5300-b836-11eb-81d1-4b7d49b5f4dd.png)

nemesisdesign reviewed (changes requested):
Great progress @niteshsinha17!

The gray background on active menu group and links does not work well.
We should try something different. Maybe we can make it closer to white (#fff) and the text color can be darker, closer to black (#000).

![Screenshot from 2021-05-22 14-26-30](https://user-images.githubusercontent.com/841044/119239100-71e88e80-bb0c-11eb-960a-e48b57b37471.png)

Regarding the bar above, which is always visible, I think on wide screens it's kind of unnecessary, maybe we can find a way to avoid it in the first place?
Did you notice in django-baton how they put the username below the logo?

![Screenshot from 2021-05-22 14-46-25](https://user-images.githubusercontent.com/841044/119239123-a6f4e100-bb0c-11eb-8861-89e0f9391810.png)

Maybe we can do something similar, move the user icon and username below the logo.
That would allow us to remove that bar entirely on wide screens.
It would be nice to still allow the user to minimize the menu but without occupying a whole bar for a menu button, 

On narrow screen, I think we can keep the bar in fixed position on top, but it would be nice to show the logo on the bar (maybe we can allow showing a smaller version, eg: square version like here on github) and when the menu is opened it may be better
to let it take the whole screen, while also make it easier to close it (the x seems a bit small, I would also make sure the menu closes itself if anything else but links are clicked/tapped).

Let's look at how github does it:

![Screenshot from 2021-05-22 14-48-42](https://user-images.githubusercontent.com/841044/119239186-14087680-bb0d-11eb-9fbf-91083069cec1.png)
![Screenshot from 2021-05-22 14-49-36](https://user-images.githubusercontent.com/841044/119239188-166ad080-bb0d-11eb-9efe-d26d9e811391.png)

We can look at popular web sites to see how they solve these issues and borrow the best practices.

Pay attention to the size of the headings, ensure they're consistent on different screen sizes.

![Screenshot from 2021-05-22 14-43-20](https://user-images.githubusercontent.com/841044/119239203-326e7200-bb0d-11eb-9da8-53e7402052fc.png)

niteshsinha17 commented:
@nemesisdesign I have few questions
>That would allow us to remove that bar entirely on wide screens.
It would be nice to still allow the user to minimize the menu but without occupying a whole bar for a menu button

We need to show a notification icon also here. What do you think about this?

>On narrow screen, I think we can keep the bar in fixed position on top, but it would be nice to show the logo on the bar (maybe we can allow showing a smaller version, eg: square version like here on github) and when the menu is opened it may be better
to let it take the whole screen, while also make it easier to close it (the x seems a bit small, I would also make sure the menu closes itself if anything else but links are clicked/tapped).

Do you want to change the design for this? should we also change opening direction of menu like git have? 
And do we need to hide logo which is inside the menu?

Btw we were following this design https://coreui.io/demo/free/3.4.0/.

nemesisdesign reviewed (commented):
> @nemesisdesign I have few questions
> 
> > That would allow us to remove that bar entirely on wide screens.
> > It would be nice to still allow the user to minimize the menu but without occupying a whole bar for a menu button
> 
> We need to show a notification icon also here. What do you think about this?

Yes, I was completely forgetting about this!
 
> > On narrow screen, I think we can keep the bar in fixed position on top, but it would be nice to show the logo on the bar (maybe we can allow showing a smaller version, eg: square version like here on github) and when the menu is opened it may be better
> > to let it take the whole screen, while also make it easier to close it (the x seems a bit small, I would also make sure the menu closes itself if anything else but links are clicked/tapped).
> 
> Do you want to change the design for this? should we also change opening direction of menu like git have?
> And do we need to hide logo which is inside the menu?
> 
> Btw we were following this design https://coreui.io/demo/free/3.4.0/.

I think we could do another iteration, also because we had forgotten about the notifications widget.

On mobile we could simply fade in the menu but I have no strong opinions, whatever works nicely will be good.

If we show a square version of the logo on narrow screens, we may show the full version in the menu, that'd be nice :blush:.

niteshsinha17 commented:
Sure, I will be creating its design soon before writing any code.

niteshsinha17 commented:
@nemesisdesign, @atb00ker and @kkreitmair I have updated this PR. Added some dummy elements in the menu so that you can test which will be removed before merging.

atb00ker reviewed (approved):
LGTM! :smile:
@niteshsinha17 can we also see if it's possible to change the background color of the svg images generated for charts on the index page?

niteshsinha17 commented:
> LGTM! 
> @niteshsinha17 can we also see if it's possible to change the background color of the svg images generated for charts on the index page?

I will check and let you know

kkreitmair commented on openwisp_utils/admin_theme/static/admin/js/menu.js:
```diff
@@ -1,21 +1,52 @@
-    setHeightToMenu();
-    window.addEventListener('resize', setHeightToMenu);
-  });
+const owContainer = document.getElementById('container');
+const owMenu = document.getElementById("menu");
+const owMainContent = document.getElementById("main-content");
+const owMenuToggle = document.querySelector('.menu-toggle');
```
Please check if the vars are not `null` when you want to work with them like in line 29, 33 ...
There are cases, where no Menu will be rendered.

The exceptions are not dramatic, but will pollute the console log unnecessarily.

kkreitmair reviewed (commented):
@niteshsinha17 Nice redesign!

As @atb00ker already mentioned for the graphs, due to the background color change, which is good in my opinion, some elements needs further adjustment.
For example the missing background of this changelist view element for operators:
![image](https://user-images.githubusercontent.com/37412542/122875546-b39b6d80-d334-11eb-8715-e7b16c3a9134.png)

Also the contrast of some elements are lower through this change for my favor.
For example the contrast of the submit row is not really visible to me:
![image](https://user-images.githubusercontent.com/37412542/122878240-d0857000-d337-11eb-919a-24b14eb00bdb.png)
See also the "Zebra striped table", "Resent Actions" section on the home page and look around for other similar elements.

atb00ker commented:
> due to the background color change, which is good in my opinion, some elements needs further adjustment.

Maybe it would be easier to go back to the white color instead of testing and fixing them in each page, what do you think @niteshsinha17? 

>  contrast of the submit row is not really visible to me

Hey, when modifing colors, are you checking for accessibility?
This might help with accessibility issues & deciding colors: 
1. https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/
2. Google chrome has axe and something called "lighthouse" in dev tools panel, you can check that out as well! :smile:

niteshsinha17 commented:
>Maybe it would be easier to go back to the white color instead of testing and fixing them in each page, what do you think @niteshsinha17?

I was also thinking about it. We may need to modify buttons too.

atb00ker commented:
Found a new bug:

![bugmore](https://user-images.githubusercontent.com/24985760/123524587-0ab09200-d6e9-11eb-900e-350047f38802.png)

The 3 burger menu icon comes over breadcrumbs in object list view when we scroll.

atb00ker commented:
Same problem on index page:

![bugmore](https://user-images.githubusercontent.com/24985760/123524669-8ca0bb00-d6e9-11eb-8959-719350c2dbc3.png)

nemesisdesign reviewed (changes requested):
Great progress @niteshsinha17!

Here's the issues I found during testing.

PS: it's not exhaustive, I had to stop because my internet went off and had to use mobile data to upload the screenshot and submit the review.

### The breadcrumb text is partly white on white background.

![Screenshot from 2021-07-04 20-45-04](https://user-images.githubusercontent.com/841044/124407126-c5075100-dd08-11eb-933c-b497c7f5c175.png)
![Screenshot from 2021-07-04 20-45-15](https://user-images.githubusercontent.com/841044/124407130-c59fe780-dd08-11eb-89a3-a1fffb376826.png)

Please make sure non-link text is of the same color of the body text and make sure links are underlined, coloured (aren't we going to use red) and make sure there's an hover effect.

We should use the same style for links on other parts of the theme BTW.

### Menu elements repeated

![Screenshot from 2021-07-04 20-46-51](https://user-images.githubusercontent.com/841044/124407234-ff70ee00-dd08-11eb-9832-d6025dd33f89.png)

Is this just for testing or what?

### Account dropdown whitespace a bit off

![Screenshot from 2021-07-04 20-47-54](https://user-images.githubusercontent.com/841044/124407298-28917e80-dd09-11eb-97bc-6d191621d43c.png)

Probably would be better to have the account menu not so pegged to the right, probably the same space as the history button there?

### Account menu items hover effect

Can you improve a bit the hover effect of the account menu items so that the text is not underline but we use a change of background instead?
It could be similar to the hover effect of the main menu.

It would be also great to have a hover effect on the first button (the one with the username shown)

### Scrollbar on minimized menu

![Screenshot from 2021-07-04 20-49-23](https://user-images.githubusercontent.com/841044/124407417-773f1880-dd09-11eb-8ac2-4229e45e588f.png)

Is this intentional? It's weird.

Maybe we could change the behavior of the menu when minimized so that it scrolls with the rest of the page and doesn't need its own scrollbar?

niteshsinha17 commented:
>Is this just for testing or what?

Yes, It's for testing. Should I remove it? BTW they will not come when you will use it with other modules. We can remove it before merging. I have already added two TODO for it.

atb00ker reviewed (commented):
The scrollbar in the menu makes this button look rather weird, can we do something about it?
![ss1](https://user-images.githubusercontent.com/24985760/124484467-10813400-ddc9-11eb-96ca-be122fd8149a.png)

The password icon seems to be cut.
![ss2](https://user-images.githubusercontent.com/24985760/124484475-124af780-ddc9-11eb-96a6-d6c14adc9985.png)

niteshsinha17 commented:
>The scrollbar in the menu makes this button look rather weird, can we do something about it?

I think when we will style the scrollbar it may not look so weird.

nemesisdesign reviewed (commented):
> > The scrollbar in the menu makes this button look rather weird, can we do something about it?
> 
> I think when we will style the scrollbar it may not look so weird.

@atb00ker @atb00ker what about just letting the menu on the left cause the main body to scroll? We should not have short pages and very long menu anyway.

niteshsinha17 commented:
Can we use this to close the menu?

![image](https://user-images.githubusercontent.com/42765292/124540218-9094b180-de3c-11eb-8970-6baaf569c107.png)

atb00ker commented:
> @atb00ker what about just letting the menu on the left cause the main body to scroll? We should not have short pages and very long menu anyway.

That sounds good as well! :smile:

pniaps commented:
I think that sidebar with it's own scrollbar is better than with the full page, I see that in almost all admin panels.

> The scrollbar in the menu makes this button look rather weird, can we do something about it?
> ![ss1](https://user-images.githubusercontent.com/24985760/124484467-10813400-ddc9-11eb-96ca-be122fd8149a.png)

Collapse-expand menu should be done with the same toggle button. It's a pain click a button to collapse and go with mouse to other button (maybe far away) to expand it.

nickberry17 commented on openwisp_utils/admin_theme/static/admin/css/ow-dashboard.css:
```diff
@@ -1,5 +1,5 @@
 #plot-container {
-  padding: 0 40px 10px 40px;
+  padding: 10px 40px;
```
I would encourage you to experiment with other CSS units, such as `fr`, or tie it to the content's text size by using `rem` instead of hardcoding it with `px`.

niteshsinha17 replied:
@nickberry17 Should I use it everywhere?

nickberry17 replied:
To do that the style and probably the HTML would need to be refactored.  However, judicious use of the new CSS features like `grid`, `flexbox`, `clamp()` and `calc()` could probably result in a more extensible and maintainable result with far less code.  Semi-related, I am seeing `!important` sprinkled about in the other CSS files which indicates that the structure of the CSS should probably be thought about again.  Check YouTube for some[ CSS best practices 2021](https://www.youtube.com/results?search_query=css+best+practices+2021) and you will see what I mean.  Google puts out some videos, and there are others.

nickberry17 commented on openwisp_utils/admin_theme/static/admin/js/menu.js:
```diff
@@ -1,21 +1,118 @@
+    // Transition fix: Add transition to menu and main content
+    // after some time.
+    if (owMenu) {
+      owMenu.style.transitionDuration = '0.3s';
```
> 0.3s

I would prefer 0.2s or 0.1s to make it feel snappier.  Also, since there may be different preferences, you could define this `transition-duration` as a variable.

nickberry17 commented on tests/test_project/apps.py:
```diff
@@ -120,21 +119,36 @@ def register_menu_groups(self):
         register_menu_group(position=3, config=docs_config)
+
+        # TODO: remove everything below it before merging
+        register_menu_group(position=4, config=auth_config)
```
Can this be done inside a `for` loop?

niteshsinha17 replied:
Yes but I will removed soon.

nickberry17 commented:
> When this is merged, let's also look at the scrollbars and make them modern to make them feel the same as the new design:
> We will need to test them with Chrome & Firefox!
> Please pick any design you like here is just a sample list: https://www.cssscript.com/best-custom-scrollbar-javascript-libraries/
> 
> ![scrollbar1](https://user-images.githubusercontent.com/24985760/117536333-14d2e000-b018-11eb-8486-92117fc3a24c.png)
> ![scrollbar2](https://user-images.githubusercontent.com/24985760/117536336-169ca380-b018-11eb-846a-1b9f8f178b59.png)

This may be a good starting point for customising the scrollbar. 

[https://www.digitalocean.com/community/tutorials/css-scrollbars](https://www.digitalocean.com/community/tutorials/css-scrollbars)

atb00ker commented on openwisp_utils/admin_theme/static/ui/openwisp/images/replies.svg:
```diff
@@ -0,0 +1,4 @@
+<svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 0 24 24" width="24px" fill="#000000">
+  <path d="M0 0h24v24H0z" fill="none"/>
+  <path d="M10 9V5l-7 7 7 7v-4.1c5 0 8.5 1.6 11 5.1-1-5-4-10-11-11z"/>
+</svg>
```
End line missing.

atb00ker reviewed (commented):
Two comments, the latest design looks good to me! :smile:

atb00ker reviewed (commented):
One more thing, when I click on menu items, they get underlined.
The underlining doesn't look good, can we remove it?

Also, section titles like "Sites", "Testapp", "Users and orgs" are also underlined, can we change their styles?

atb00ker reviewed (changes requested):
Found one more problem, the silding animation on drawer closing is busted.
The content flies from the right in my screen, looks odd!

atb00ker reviewed (commented):
Also, can you please check what happens when we add alot of times in the submenu of the last time, does it spill outside the page? (Haven't tested it, but thought about it after exiting the review, so just check it once please!)

niteshsinha17 commented:
> busted

@atb00ker Can you please share a screenshot or gif. I am not getting what do you mean

niteshsinha17 commented:
> Also, can you please check what happens when we add alot of times in the submenu of the last time, does it spill outside the page? (Haven't tested it, but thought about it after exiting the review, so just check it once please!)

It looks like this 
![image](https://user-images.githubusercontent.com/42765292/125791118-5b3aa27f-87f7-484a-bc7c-ef800425bbc8.png)

niteshsinha17 commented:
> One more thing, when I click on menu items, they get underlined.
> The underlining doesn't look good, can we remove it?
> 
> Also, section titles like "Sites", "Testapp", "Users and orgs" are also underlined, can we change their styles?

It was recommended by @nemesisdesign to keep links the same as we have here https://openwisp.org/

Do we need to change it?

atb00ker commented:
> @atb00ker I got what do you mean. It's because I am hiding it when the menu changes state. We can't change this but can reduce by decreasing the closing time.

Then we should change the opening / closing logic entirely, maybe when the menu is open it comes over the rest of the content without displacing it; that should ensure this doesn't happen, that said, I really think it's not a hard blocker, can you please spend some time thinking about how to remove this problem with minimal efforts?

atb00ker reviewed (commented):
I found 2 more (potential) problems: 

1. The change icon seems to be missing, is it me again?:

![iconMissing](https://user-images.githubusercontent.com/24985760/125831977-47d5ceef-99d0-42b0-8147-15046feff7d2.png)


2. I clicked on logout and it threw an error on me, is it my local setup or are you able to reproduce it?
![errorTwo](https://user-images.githubusercontent.com/24985760/125831984-b5b51256-2d77-4152-8d22-8e94ee9d46df.png)

kkreitmair commented:
Two points from my side:

On small screens, the account button gets to small, so that the user name is not visible anymore.
![image](https://user-images.githubusercontent.com/37412542/125990187-83c6c4b6-3cb2-4e82-9207-493190096338.png)

Also on small screens, the menu not overlaps the scrollbar of the list view behind. (I tested it with the user listview)
![image](https://user-images.githubusercontent.com/37412542/125990233-2d0c4bb5-f9f3-4948-927f-35f94b30793b.png)

All in all, the design is great. Keep going this way.

niteshsinha17 commented:
@kkreitmair Can you please pull again and check this once more because I am not able to reproduce these issues. Everything looks fine on my screen. You may clear the cache and try again.

nemesisdesign reviewed (changes requested):
It's looking really great!

Here's some areas which I think can be refined. This is not exhaustive, I plan on doing more testing and leaving more feedback asap.

Definitely going in the right direction though! :+1: :clap:
Let's keep pushing! :muscle: :rocket: :1st_place_medal:  

#### Side button to minimize menu is big

![minimize-menu](https://user-images.githubusercontent.com/841044/126050784-9d7c340e-d6d9-42b9-9271-2d736e590c3f.gif)

I think this occupying quite a bit of space there and it could be clicked often by accident.
I think it could be made a bit more subtle and less distracting. 

#### Main site scrollbar

![Screenshot from 2021-07-17 17-40-36](https://user-images.githubusercontent.com/841044/126050834-196f0011-c9e0-46bb-860f-b371538c9514.png)

The main page scrollbar suddenly become very slim, is that on purpose?
I don't think we should make the main page scrollbar so slim.
I can't even grab it with the mouse. I personally don't care about this because I use the wheel, trackpad, keyboard to scroll, but I have seen some people grabbing the scrollbar regularly :scream: and this won't help them.

#### Minimized menu disappears :grin: 

It seems I gave a not so good suggestion!

Look at what happens at the menu when minimizing on a long page:

![minimized-menu-issue](https://user-images.githubusercontent.com/841044/126051639-045e48e9-2d92-475c-9ec5-e4f6ade8dc3f.gif)

And puff, it's like magic :joy: !

Not the best thing, right?

Is it possible to allow the minimized version of the menu to scroll without showing the scrollbar? Or is it possible to show the scrollbar only while the scrolling is taking place?

niteshsinha17 commented:
>If we go back to fixed position, Is it possible to allow the minimized version of the menu to scroll without showing the scrollbar? Or is it possible to show the scrollbar only while the scrolling is taking place?

@nemesisdesign We can do this too but again we will come back to the old state that we had discussed. More js code will be needed then.

What if the menu toggle button always remains in the same position even we scroll like this.

![scroll](https://user-images.githubusercontent.com/42765292/126067929-e1c57f4a-ac5e-4bbd-b991-09443200ffc3.gif)

Please notice the width of the button and scrollbar also. Does it look fine?

kkreitmair commented:
> @kkreitmair Can you please pull again and check this once more because I am not able to reproduce these issues. Everything looks fine on my screen. You may clear the cache and try again.

@niteshsinha17 I checked it again. The account button is now fine.
The Problem with the scrollbar still exists, but only in Firefox "Responsive Design View" :see_no_evil:.
![Peek 2021-07-19 09-29](https://user-images.githubusercontent.com/37412542/126121446-ec5559c3-df0b-4696-9af2-e1af19c6b545.gif)

Mediaquerries are not changing between this views, so i would say, it is not a problem on our side.

nemesisdesign reviewed (commented):
Can you please pull this https://github.com/openwisp/openwisp-utils/commit/d559765f323d692dcaa4dceac6c4440aa507239b

niteshsinha17 commented:
> Can you please pull this [d559765](https://github.com/openwisp/openwisp-utils/commit/d559765f323d692dcaa4dceac6c4440aa507239b)

@nemesisdesign Done
@kkreitmair Can you please check again the issue you mentioned here https://github.com/openwisp/openwisp-utils/pull/185#issuecomment-882332926  I have fixed it

nemesisdesign reviewed (changes requested):
Great progress @niteshsinha17, here's the issue I found during this round of testing.

### Lateral scrollbar hidden

![lateral-scrollbar-issue](https://user-images.githubusercontent.com/841044/126581102-d0307530-227c-4eaf-bb9b-6df38649de2f.gif)

Seems that the lateral scrollbar is hidden for good, why is that? This shouldn't happen.

### Active menu item not shown

![Screenshot from 2021-07-21 21-05-14](https://user-images.githubusercontent.com/841044/126581222-e0732c0d-616a-4eba-9435-089b7b0ef90d.png)

I am in the user page here but no indication is given in the menu. Would it be possible to highlight the group? Is that hard or can it be done?

niteshsinha17 commented:
>I am in the user page here but no indication is given in the menu. Would it be possible to highlight the group? Is that hard or can it be done?

I think it is possible and we will need to identify the group according to the current URL for the page and another piece of JS code will be required for this. This PR already contains a lot of work. It will be easy to work on it separately when this PR is done.
What do you think?

nemesisdesign commented:
> > I am in the user page here but no indication is given in the menu. Would it be possible to highlight the group? Is that hard or can it be done?
> 
> I think it is possible and we will need to identify the group according to the current URL for the page and another piece of JS code will be required for this. This PR already contains a lot of work. It will be easy to work on it separately when this PR is done.
> What do you think?

Fair enough: https://github.com/openwisp/openwisp-utils/issues/216

> > Seems that the lateral scrollbar is hidden for good, why is that? This shouldn't happen.
> 
> @nemesisdesign This was also surprising for me when I saw it today. I have found the issue. It's happening only on the change list page because of filters. I think we can ignore it for now as we are going to change filters too. I have checked when we remove filters, the page works fine.

I insist that we fix this now because I want to start deploying this in real systems, because again I stress that the only way to improve software is to use it in the real world, if we don't do this asap, this risks to be nice work which won't be used. If GSoC finishes and we don't catch all major problems before then, this risks to never be released, so we have to address any major issue like this now, deploy it, use it, stress test it further.

niteshsinha17 commented on openwisp_utils/admin_theme/static/admin/css/openwisp.css:
```diff
@@ -792,6 +777,7 @@ h1#site-name {
   #container {
     overflow: scroll;
     height: 100vh;
+    scrollbar-width: none; /* firefox fix*/
```
@kkreitmair I have added this line for the fix.

atb00ker reviewed (commented):
1. Scrollbars are working on chrome / chromium but failing on FF. (Tried version 78esr and 89)
2. Can we remove this underlining? (Same for Breadcrumbs)
![ss1](https://user-images.githubusercontent.com/24985760/126861262-82fabdab-2ed9-4072-9e33-1557a24e8752.png)
3. Please correct me if I am wrong, but general expectation in smaller devices is small scrollbars. (Like in `740px` screenwidth) This is very fat for `776px` screen. What do you think?
![ss2](https://user-images.githubusercontent.com/24985760/126861320-f04aff6a-e54b-4ddd-89fd-4d502a4febe6.png)
4. Why are `search` and `go` buttons in old UI styles while add has a separate style? 
![ss3](https://user-images.githubusercontent.com/24985760/126861446-8d15cb2f-ab78-4cb1-ad2c-7db48c2a7f68.png)
5. Color of these icons is out of theme.
![ss4](https://user-images.githubusercontent.com/24985760/126861516-13581a11-9a26-412e-8170-e1738c79c97d.png)
6. Why is this `+` floating the middle of nowhere. It's not clear that will help me in creating a new group.
![ss5](https://user-images.githubusercontent.com/24985760/126861598-e21c576d-d660-480b-8890-7fd8982e6f2a.png)

nemesisdesign reviewed (changes requested):
@niteshsinha17 it seems to me that in the latest version the responsiveness is kinda broken!
If I try to reduce the width of the window, the menu does not minimize automatically but stays open and covers the page.

Then if I minimize the menu and keep on fiddling with the window width, there's a point in which the menu opens by itself without me clicking on it.

![responsive-quirk](https://user-images.githubusercontent.com/841044/126908697-b271a40e-b956-4644-9268-ce4ee79b3087.gif)

nemesisdesign reviewed (changes requested):
Great work @niteshsinha17! :+1: :muscle: 

I have a considerable amount of minor improvements for readability and for implementing consistent spacing (which was a bit odd) which took me the entire day but would have taken a lot more if I had to tell you each one by one.

@niteshsinha17 @atb00ker @kkreitmair @nickberry17 please have a look and let me know if there's any doubt or anything that doesn't convince you. 

I also updated these modules:

- controller
- notifications

Please update also those ones when testing, and I suggest to test on OpenWISP Controller (with the modified OpenWISP Users branch for UX improvemens and registering menu items as well).

### Minimize menu button

I see the tooltip always says "close Menu" but this is incorrect.
Maybe we can change that to be "Toggle minimized menu"?

### Improve accessibility when browsing with keyboard TAB

Try browsing the menu using only tab (no mouse), is it possible?
Not really, which isn't good for accessibility (impaired movements, blindness, etc.) I tried to improve this a bit by making sure "focus" styles are the same as hover, which helps a bit, but still it needs more work to ensure the menu and every element which was added in this PR is browsable also with keyboard only.

Please take a look at these pages:

- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA

### Notification widget

Could you change the notification widget icon to use the same technique of masking so we can change the color there too on hover/focus please?

If you notice my changes you will see I added an hover effect to the account menu which is similar to the hover effect of the menu links.

### Menu interferes with config preview

![Screenshot from 2021-07-25 15-16-56](https://user-images.githubusercontent.com/841044/126921329-c1becb73-1a34-4dd5-ba14-da764d89e462.png)

The same happens in Advanced Mode (raw json editing of configuration data).

Please make sure the menu doesn't interfere with other elements of OpenWISP (preview, config advanced mode, etc).

### Device detail page, buttons quirk on narrow screens

![device-detail-buttons](https://user-images.githubusercontent.com/841044/126921474-ab56af2a-0a93-4aa0-a5a9-08c52940ead7.gif)

If you can find a good solution to fix the quirks above would be great!

Notice also how the send commands button is showing twice? You still can't replicate that one?

niteshsinha17 commented:
@nemesisdesign I have made changes suggested. Except the accessibility one. It will be easier for me to handle this in new PR.

nemesisdesign reviewed (commented):
Ready to merge apart from this:
https://github.com/openwisp/openwisp-controller/pull/493#pullrequestreview-717488366

nemesisdesign reviewed (approved):
We're good to go!

Let's merge and start using it to stress test it and find any potential issues before releasing it.

Thanks everyone! @niteshsinha17 @atb00ker @nickberry17 @kkreitmair @pniaps 

:+1: :clap: :muscle: :tada: :rocket:

niteshsinha17 commented:
@nemesisdesign Your last commit is missing I think.
